File: runtimes/swift/json.swift
func toJson(_ obj: Any?) -> String {
    var seen: [ObjectIdentifier] = []
    return encode(obj, &seen)
}

private func encode(_ obj: Any?, _ seen: inout [ObjectIdentifier]) -> String {
    guard let obj = obj else { return "null" }

    if let b = obj as? Bool { return b ? "true" : "false" }
    if let n = obj as? Int { return "\(n)" }
    if let n = obj as? Double {
        if n.isNaN || n.isInfinite { return "null" }
        return "\(n)" 
    }
    if let s = obj as? String { return quote(s) }
    if let d = obj as? Date { return quote(ISO8601DateFormatter().string(from: d)) }

    let ref = obj as AnyObject
    let oid = ObjectIdentifier(ref)
    guard !seen.contains(oid) else {
        fatalError("Cycle detected while encoding to json")
    }
    seen.append(oid)
    defer { seen.removeAll { $0 == oid } }

    if let dict = obj as? [String: Any?] {
        let parts = dict.map { "\(quote($0.key)):\(encode($0.value, &seen))" }
        return "{\(parts.joined(separator: ","))}"
    }

    if let arr = obj as? [Any?] {
        return "[\(arr.map { encode($0, &seen) }.joined(separator: ","))]"
    }

    let mirror = Mirror(reflecting: obj)
    if mirror.children.isEmpty {
        return quote("\(obj)")
    }
    let parts = mirror.children.compactMap { child -> String? in
        guard let label = child.label, !label.hasPrefix("_") else { return nil }
        return "\(quote(label)):\(encode(child.value, &seen))"
    }
    return "{\(parts.joined(separator: ","))}"
}

private func quote(_ s: String) -> String {
    var out = "\""
    for ch in s.unicodeScalars {
        switch ch.value {
        case 0x22: out += "\\\""
        case 0x5C: out += "\\\\"
        case 0x08: out += "\\b"
        case 0x0C: out += "\\f"
        case 0x0A: out += "\\n"
        case 0x0D: out += "\\r"
        case 0x09: out += "\\t"
        default:
            if ch.value < 0x20 {
                out += String(format: "\\u%04x", ch.value)
            } else {
                out += String(ch)
            }
        }
    }
    out += "\""
    return out
}

File: runtimes/dart/json.dart
String toJson(Object? obj) {
  final seen = <Object>{};
  final sb = StringBuffer('\n');
  sb.write(_encode(obj, seen));
  return sb.toString();
}

String _encode(Object? obj, Set<Object> seen) {
  if (obj == null) return 'null';
  if (obj is bool) return obj ? 'true' : 'false';
  if (obj is int) return obj.toString();
  if (obj is double) {
    if (obj.isNaN || obj.isInfinite) return 'null';
    return obj.toString();
  }
  if (obj is String) return _quote(obj);
  if (obj is DateTime) return _quote(obj.toIso8601String());

  if (identical(seen, obj)) throw StateError('Cycle detected while encoding to json');
  if (seen.contains(obj)) throw StateError('Cycle detected while encoding to json');
  seen.add(obj);
  try {
    if (obj is Map) {
      final parts = <String>[];
      for (final e in obj.entries) {
        final k = e.key is String ? e.key as String : e.key.toString();
        parts.add('${_quote(k)}:${_encode(e.value, seen)}');
      }
      return '{${parts.join(',')}}';
    }

    if (obj is Iterable) {
      return '[${obj.map((x) => _encode(x, seen)).join(',')}]';
    }

    try {
      final mirror = reflect(obj);
      final classMirror = mirror.type;
      final parts = <String>[];
      for (final e in classMirror.declarations.entries) {
        final decl = e.value;
        if (decl is VariableMirror && !decl.isStatic && !decl.isPrivate) {
          final name = MirrorSystem.getName(e.key);
          if (name.startsWith('_')) continue;
          final value = mirror.getField(e.key).reflectee;
          parts.add('${_quote(name)}:${_encode(value, seen)}');
        }
      }
      return '{${parts.join(',')}}';
    } catch (_) {}

    return _quote(obj.toString());
  } finally {
    seen.remove(obj);
  }
}

String _quote(String s) {
  final buf = StringBuffer('"');
  for (final ch in s.runes) {
    switch (ch) {
      case 0x22: buf.write(r'\"'); break;
      case 0x5C: buf.write(r'\\'); break;
      case 0x08: buf.write(r'\b'); break;
      case 0x0C: buf.write(r'\f'); break;
      case 0x0A: buf.write(r'\n'); break;
      case 0x0D: buf.write(r'\r'); break;
      case 0x09: buf.write(r'\t'); break;
      default:
        if (ch < 0x20) {
          buf.write('\\u${ch.toRadixString(16).padLeft(4, '0')}');
        } else {
          buf.writeCharCode(ch);
        }
    }
  }
  buf.write('"');
  return buf.toString();
}
